elements_ephys/readers/spikeglx.py

```diff
@@ -23,40 +23,76 @@ def __init__(self, root_dir):
         name & associated meta - no interpretation of g0_t0.imec, etc is
         performed at this layer.
         '''
+        self._apmeta, self._apdata = None, None
+        self._lfmeta, self._lfdata = None, None
 
-        self.root_dir = root_dir
+        self.root_dir = pathlib.Path(root_dir)
 
         meta_filepath = next(pathlib.Path(root_dir).glob('*.ap.meta'))
-        self.npx_meta = NeuropixelsMeta(meta_filepath)
-
         self.root_name = meta_filepath.name.replace('.ap.meta', '')
-        self._apdata = None
-        self._lfdata, self._lfmeta = None, None
+
+    @property
+    def apmeta(self):
+        if self._apmeta is None:
+            self._apmeta = SpikeGLXMeta(self.root_dir / (self.root_name + '.ap.meta'))
+        return self._apmeta
 
     @property
     def apdata(self):
-        if self._apdata is not None:
-            return self._apdata
-        else:
-            return self._read_bin(self.root_dir / (self.root_name + '.ap.bin'))
+        """
+        AP data: (sample x channel)
+        Channels' gains (bit_volts) applied - unit: uV
+        """
+        if self._apdata is None:
+            self._apdata = self._read_bin(self.root_dir / (self.root_name + '.ap.bin'))
+            self._apdata = self._apdata * self.get_channel_bit_volts('ap')
+        return self._apdata
 
     @property
     def lfmeta(self):
-        if self._lfmeta is not None:
-            return self._lfmeta
-        else:
-            return _read_meta(self.root_dir / (self.root_name + '.lf.meta'))
+        if self._lfmeta is None:
+            self._lfmeta = SpikeGLXMeta(self.root_dir / (self.root_name + '.lf.meta'))
+        return self._lfmeta
 
     @property
     def lfdata(self):
-        if self._lfdata is not None:
-            return self._lfdata
+        """
+        LFP data: (sample x channel)
+        Channels' gains (bit_volts) applied - unit: uV
+        """
+        if self._lfdata is None:
+            self._lfdata = self._read_bin(self.root_dir / (self.root_name + '.lf.bin'))
```

Comment:
is there a reason for separating this computation into two statements?

Reply:
I want to make it explicit that one step is to extract data from the file, and another is to apply to bit-volts conversion